dotnet/samples/KernelSyntaxExamples/Example53_FunctionHooks.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.Connectors.AI.OpenAI.Tokenizers;
+using Microsoft.SemanticKernel.SkillDefinition;
+using RepoUtils;
+
+#pragma warning disable RCS1214 // Unnecessary interpolated string.
+
+// ReSharper disable once InconsistentNaming
+public static class Example53_FunctionHooks
+{
+    public static async Task RunAsync()
+    {
+        Console.WriteLine("======== Using Function Hooks ========");
+
+        string openAIModelId = TestConfiguration.OpenAI.ModelId;
+        string openAIApiKey = TestConfiguration.OpenAI.ApiKey;
+
+        if (openAIModelId == null || openAIApiKey == null)
+        {
+            Console.WriteLine("OpenAI credentials not found. Skipping example.");
+            return;
+        }
+
+        IKernel kernel = new KernelBuilder()
+            .WithLogger(ConsoleLogger.Logger)
+            .WithOpenAITextCompletionService(
+                modelId: openAIModelId,
+                apiKey: openAIApiKey)
+            .Build();
+
+        const string FunctionDefinition = "Write a paragraph about: {{$input}}.";
+
+        var excuseFunction = kernel.CreateSemanticFunction(FunctionDefinition, maxTokens: 100, temperature: 0.4, topP: 1);
+
+        Task MyPreHook(PreExecutionContext executionContext)
+        {
+            Console.WriteLine($"Pre Hook - Prompt: {executionContext.Prompt}");
+
+            return Task.CompletedTask;
+        }
+
+        Task MyCancelledPreHook(PreExecutionContext executionContext)
+        {
+            Console.WriteLine("Pre Hook - Should not trigger");
+
+            return Task.CompletedTask;
+        }
+
+        Task MyPreHook2(PreExecutionContext executionContext)
+        {
+            Console.WriteLine($"Pre Hook 2 - Prompt Token: {GPT3Tokenizer.Encode(executionContext.Prompt!).Count}");
+
+            return Task.CompletedTask;
+        }
+
+        Task MyPostHook(PostExecutionContext executionContext)
+        {
+            Console.WriteLine($"Post Hook - Total Tokens: {executionContext.SKContext.ModelResults.First().GetOpenAITextResult().Usage.TotalTokens}");
+            return Task.CompletedTask;
+        }
+
+        excuseFunction.SetPreExecutionHook(MyPreHook);
```

Comment:
What if my semantic function calls other semantic functions? Will the token usage be cumulative across everything in the three?

Reply:
@lemillermicrosoft In the PreExecutionHook the generated prompt will be the version that the template engine generates at that stage which by definition will be cumulative of all dependencies the template engine has to generate the prompt.